### ISSUE TITLE
Could not load type 'System.IdentityModel.Tokens.TokenValidationParameters' 

### ISSUE THREAD

**Issue by bluewalk** [jakubgarfield/Bonobo-Git-Server#681]
Hi,

I tried enabling ADFS according to the instructions found at https://bonobogitserver.com/federation-authentication/. All is well, untill I hit F5:
![2017-04-25 15_30_57-could not load type system identitymodel tokens tokenvalidationparameters from](https://cloud.githubusercontent.com/assets/6522477/25387762/3d4564d0-29cc-11e7-92a8-f6a612351da1.png)

Running latest version 6.1.0 although the binaries have version 6.0.0.556

**Comment by willdean** [comment]
@bluewalk Thanks for this.  I think the `system.identitymodel` namespace/libraries were part of some evil alien plot to reduce happiness among Earthlings.

We'll have a look at what's happening.

**Comment by bluewalk** [comment]
I'm pretty sure thats what's happening, I know they're very, and I mean very, version specific combined with Owin.
Looking forward to the solution :)

**Comment by willdean** [comment]
@bluewalk Yes, I've just had the misery of doing a jwt nuget update on another project.  Eventually I just backed the whole darn thing out because it was more hassle than I could bear.

**Comment by bluewalk** [comment]
Just checked out the source and fiddled around with it, downgrading JWT solves the problem:

```
Install-Package System.IdentityModel.Tokens.Jwt -Version 4.0.2.206221351
```

As per https://github.com/IdentityServer/IdentityServer3/issues/3017

**Comment by willdean** [comment]
@bluewalk Thanks for this about the 4.x/5.0 issue - we'll downgrade the jwt library and release a new version.  It does seem lots of people have been caught by this - it's really poor that nuget offers an upgrade to an incompatible version, but I guess it fits well with the whole core shambles.

I don't know much about ADFS, but I know we changed the specific claims which Bonobo uses to make the mapping more intuitive, but I think it might always be necessary to have some claims mapping in ADFS.

The last person who helped us with ADFS testing disappeared and so we never really got to the bottom of if it is as good as it could be - if you're able to help test things then we can try and improve it.

**Comment by bluewalk** [comment]
I'm available for ADFS testing for sure!
Yes, from ADFS you will get roles as well (which will basically be AD groups). After login these need to be matched with the ones given in web.config and assign the corresponding roles.
I will need to get into the code and the flow to see where this process can take place. If you can point me in the right direction I can have a look as well :-)

**Comment by willdean** [comment]
@bluewalk When you have a moment (and when it's finished building), would you mind checking if the artifacts from the test build in PR #689 load up the Jwt library OK for you?

I'm going to look at the roles issue separately.

**Comment by willdean** [comment]
@bluewalk Would you mind having a look through #368 (sorry, there's a lot of it).

There's a user there who got a complete ADFS setup working, and described what he/she set up to do it (he/she has some additional rules around group membership which may solve your admin issue.

If you're happy that's a useful set of instructions, then perhaps you'd help us get them into the documentation properly?  @larshg  and I are a bit out of our depth with ADFS and neither of us has a setup to test against.

**Comment by bluewalk** [comment]
@willdean Yes the Jwt library loads up normally now. Will have a look at #368 and once I have it working I will be happy to help adding it to the documentation.

**Comment by bluewalk** [comment]
@willdean Okay got it working. The documentation at https://bonobogitserver.com/federation-authentication/ should be altered, first the `How to configure Bonobo Git Server?` should move to the top and then the contents for `How to configure ADFS?` need to contain the following:

1. Open AD FS Management
2. Click `Add Relying Party Trust`
3. Choose `Claims aware` and click `Start`
4. Choose `Enter data about the relaying party manually` and click `Next`
5. Enter a display name and a description if you want to and click `Next`
6. We can skip this page so just click `Next`
7. Choose `Enable support for the WS-Federation Passive protocol` and enter the URL to your git server (must be https)
8. Add the contents of your web.config `FederationRealm` to `Relying party trust identifier` if it differs from your url and click `Next`
9. Now just click `Next` and `Next`, leave `Configure claims issuance policy for this application` checked and then `Close`

Now we are at the `Edit Claims Issuance Policy for **your name**`, we have to add two rules here.
1. Click `Add Rule`
2. Choose `Send Claims Using a Custom Rule` and click `Next`
3. Enter a `Claim rule name`, use `Active Directory` for example
4. Paste the following in `Custom rule`: 
`c:[Type == "http://schemas.microsoft.com/ws/2008/06/identity/claims/windowsaccountname", Issuer == "AD AUTHORITY"]
 => issue(store = "Active Directory", types = ("http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname", "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname", "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress", "http://schemas.microsoft.com/ws/2008/06/identity/claims/role", "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn", "http://temp.org/identity/claims/adObjectGuidBase64org"), query = ";givenName,sn,mail,tokenGroups,memberOf,ObjectGuid;{0}", param = c.Value);`
5. Click `Finish`

And the second rule:
1. Click `Add Rule`
2. Choose `Send Claims Using a Custom Rule` and click `Next`
3. Enter a `Claim rule name`, use `Name ID` for example
4. Paste the following in `Custom rule`: 
`c:[Type == "http://temp.org/identity/claims/adObjectGuidBase64org"]
 => issue(Type = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier", Issuer = c.Issuer, OriginalIssuer = c.OriginalIssuer, Value = c.Value, ValueType = c.ValueType, Properties["http://schemas.xmlsoap.org/ws/2005/05/identity/claimproperties/format"] = "urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified");`
5. Click `Finish`

All we have to do now is click `Close` and AD FS has been setup correctly. Now configure Bonobo Git Server for ADFS

**Comment by willdean** [comment]
@bluewalk Many thanks for this - I have put into into the Federation Auth page on the website, but I have not yet updated the website.

Would you mind having a look at the page [here](https://github.com/jakubgarfield/Bonobo-Git-Server-Web/blob/master/content/federation-authentication.md)

If there's anything you want to change there, I think you may be able to do an edit directly in GH and have it generate a PR.

**Comment by bluewalk** [comment]
@willdean I have created a pull request, there was just one incorrect statement which I've removed. Other than that it looks great!

**Comment by willdean** [comment]
@bluewalk That's great - thanks - I've just merged that, and it will get published when we release 6.2.

Many thanks for your help with this - I'm going to close this issue now.
